test/functional/mempool_expiry.py

```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/env python3
+# Copyright (c) 2020 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Tests that a mempool transaction expires after a given timeout and that its
+children are removed as well.
+
+Both the default expiry timeout defined by DEFAULT_MEMPOOL_EXPIRY and a user
+definable expiry timeout via the '-mempoolexpiry=<n>' command line argument
+(<n> is the timeout in hours) are tested.
+"""
+
+from datetime import timedelta
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import (
+    assert_equal,
+    assert_raises_rpc_error,
+    find_vout_for_address,
+)
+
+DEFAULT_MEMPOOL_EXPIRY = 336
+CUSTOM_MEMPOOL_EXPIRY = 10
+
+
+class MempoolExpiryTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 1
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+
+    def test_transaction_expiry(self, timeout):
+        """Tests that a transaction expires after the expiry timeout and its
+        children are removed as well."""
+        node = self.nodes[0]
+
+        # Send a parent transaction that will expire.
+        parent_address = node.getnewaddress()
+        parent_txid = node.sendtoaddress(parent_address, 1.0)
+
+        # Set the mocktime to the arrival time of the parent transaction.
+        entry_time = node.getmempoolentry(parent_txid)["time"]
+        node.setmocktime(entry_time)
+
+        # Create child transaction spending the parent transaction
+        vout = find_vout_for_address(node, parent_txid, parent_address)
+        inputs = [{'txid': parent_txid, 'vout': vout}]
+        outputs = {node.getnewaddress(): 0.99}
+        child_raw = node.createrawtransaction(inputs, outputs)
+        child_signed = node.signrawtransactionwithwallet(child_raw)["hex"]
+
+        # Let half of the timeout elapse and broadcast the child transaction.
+        half_expiry_time = entry_time + int(60 * 60 * timeout/2)
+        node.setmocktime(half_expiry_time)
+        child_txid = node.sendrawtransaction(child_signed)
+        self.log.info("Broadcast child transaction after {} hours.".format(
+            str(timedelta(seconds=(half_expiry_time-entry_time)))))
+
+        # Let most of the timeout elapse and check that the parent tx is still
+        # in the mempool.
+        nearly_expiry_time = entry_time + 60 * 60 * timeout - 5
+        node.setmocktime(nearly_expiry_time)
+        # Expiry of mempool transactions is only checked when a new transaction
+        # is added to the to the mempool.
+        node.sendtoaddress(node.getnewaddress(), 1.0)
+        self.log.info("Test parent tx not expired after {} hours.".format(
```

Comment:
nit, maybe keep these as comment instead? Not sure if logging this adds value.

Reply:
Logging is preferred over inline comments in the test because logging statements serve as "checkpoints" (to see how far a test run until failure). Also, they help to illustrate the test log better, which would otherwise only contain Bitcoin Core logs.

Reply:
When a test fails we see trace. I think logging makes sense when it shows some computed values or so, which a trace doesn't help.

Reply:
The trace is one data point that can help debug a test failure. The entire history of the log up to that point is also useful. For example transaction relay (inv-getdata) often takes several seconds, so it is useful to know when a subtest has started and when it ended. If some Bitcoin Core logs from one subtest interleave with the ones of another subtest we can conclude that we are missing a syncronisation call. This is the most common test error we are running into.

Reply:
Yes makes sense in those cases.